torch/distributed/_sharded_tensor/__init__.py

```diff
@@ -78,3 +82,39 @@ def init_from_local_shards(
         local_shards,
         sharded_tensor_metadata,
         process_group=process_group)
+
+def state_dict_hook(module, destination, prefix, local_metadata):
+    """
+    Hook to add ShardedTensor to Module's ``state_dict``. Needs to be
+    registered to the Module using
+    :meth:`torch.nn.Module._register_state_dict_hook`.
+    """
+    _recurse_update_dict(module, destination, prefix)
+
+def pre_load_state_dict_hook(module, state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs):
+    """
+    Pre-load state dict hook to add ShardedTensor to the module.
+    """
+    _recurse_update_module(module, state_dict, prefix)
+
+def _recurse_update_module(module, state_dict, prefix):
+    for attr_name, attr in module.__dict__.items():
+        key = prefix + attr_name
+        if key in state_dict:
+            if isinstance(state_dict[key], ShardedTensor):
+                setattr(module, attr_name, state_dict[key])
+
+    for submodule_name, submodule in module.named_modules():
+        key = prefix + submodule_name
+        if submodule_name:
+            _recurse_update_module(submodule, state_dict, key + '.')
+
+
+def _recurse_update_dict(module, destination, prefix):
+    for attr_name, attr in module.__dict__.items():
+        if isinstance(attr, ShardedTensor):
+            destination[prefix + attr_name] = attr
+
+    for submodule_name, submodule in module.named_modules():
+        if submodule_name != '':
+            _recurse_update_dict(submodule, destination, prefix + submodule_name + '.')
```

Comment:
I posted an issue here #68805 for a potential improvement. I believe the recursion here is not necessary and causes inefficiency when retrieving the state dict.